flink-scala-shell/src/main/java/org/apache/flink/api/java/ScalaShellRemoteEnvironment.java

```diff
@@ -19,51 +19,101 @@
  * limitations under the License.
  */
 
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.api.common.InvalidProgramException;
 import org.apache.flink.api.common.JobExecutionResult;
 import org.apache.flink.api.common.Plan;
 import org.apache.flink.api.common.PlanExecutor;
 import org.apache.flink.api.scala.FlinkILoop;
 import org.apache.flink.configuration.Configuration;
+import org.apache.flink.util.JarUtils;
+import org.apache.flink.util.function.TriFunction;
 
+import java.io.File;
+import java.io.IOException;
 import java.net.MalformedURLException;
 import java.net.URL;
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.List;
 
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
 /**
- * Special version of {@link org.apache.flink.api.java.RemoteEnvironment} that has a reference
- * to a {@link org.apache.flink.api.scala.FlinkILoop}. When execute is called this will
- * use the reference of the ILoop to write the compiled classes of the current session to
- * a Jar file and submit these with the program.
+ * A remote {@link ExecutionEnvironment} for the Scala shell.
```

Comment:
Why we still call it `remote` ?

Reply:
It still connects to a remote cluster from within scala shell

Reply:
In semantic perspective, it is weird to call it remote ExecutionEnvironment, but didn't extend it `RemoteEnvironment`. Could you explain more about the motivation ?

Reply:
Technically the configuration code path of `RemoteEnvironment` is different from `ScalaShellRemoteEnvironment`. Duplicate codes are hidden back `PlanExecutor`. If you call current `RemoteEnvironment` as its real usage `LocalRemoteEnvironment` you will find that `LocalRemoteEnvironment` is at the same level of `ScalaShellRemoteEnvironment`. No inheritance here.

Reply:
Technically the configuration code path of `RemoteEnvironment` is different from `ScalaShellRemoteEnvironment`. Duplicate codes are hidden back PlanExecutor. But I think you're right that we can make this configuration flow configuration(appendable). Hopefully I send a new pull request integrate this thought.